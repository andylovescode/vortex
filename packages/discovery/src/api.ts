export type DiscoveryTarget = "client" | "server";
export interface DiscoveryPluginProps {
	target: DiscoveryTarget;
}
export interface DiscoveryProps extends DiscoveryPluginProps {
	source: string;
	jsx: boolean;
	typescript: boolean;
	fileName: string;
}
export type Discovery = {
	type: "route_frame";
	path: string;
	exported: string;
	frameType: "page" | "layout";
};
