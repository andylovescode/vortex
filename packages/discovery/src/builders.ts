import type {
	Expression,
	LabelIdentifier,
	ModuleDeclaration,
	Span,
	StringLiteral,
} from "oxc-parser";
import type { CompilerState } from "./compiler";

export const defaultSpan: Span = {
	start: 0,
	end: 0,
};

export function literal<T>(value: T, span = defaultSpan): StringLiteral {
	return {
		type: "Literal",
		value,
		...span,
	} as unknown as StringLiteral;
}

export function exportNode(state: CompilerState, node: Expression): string {
	const transformer = state.transformer;
	const id = transformer.getExportId();

	const decl: ModuleDeclaration = {
		type: "ExportNamedDeclaration",
		declaration: {
			type: "VariableDeclaration",
			kind: "const",
			declarations: [
				{
					type: "VariableDeclarator",
					id: {
						type: "Identifier",
						name: id,
						...defaultSpan,
					},
					init: node,
					...defaultSpan,
				},
			],
			...defaultSpan,
		},
		specifiers: [],
		source: null,
		attributes: [],
		...defaultSpan,
	};

	transformer.addDeclaration(decl);

	return id;
}

export function identifier(name: string): LabelIdentifier {
	return {
		type: "Identifier",
		name,
		...defaultSpan,
	};
}

export function importDeclaration(
	module: string,
	name: string,
	local: string,
): ModuleDeclaration {
	return {
		type: "ImportDeclaration",
		specifiers: [
			{
				type: "ImportSpecifier",
				imported: identifier(name),
				local: identifier(local),
				...defaultSpan,
			},
		],
		source: literal(module),
		phase: null,
		importKind: "value",
		attributes: [],
		...defaultSpan,
	};
}

export function createObject(
	properties: Record<string, Expression>,
	span = defaultSpan,
): Expression & Span {
	return {
		type: "ObjectExpression",
		properties: Object.entries(properties).map(([key, value]) => ({
			type: "Property",
			key: identifier(key),
			value,
			kind: "init",
			computed: false,
			method: false,
			shorthand: false,
			...defaultSpan,
		})),
		...span,
	};
}
