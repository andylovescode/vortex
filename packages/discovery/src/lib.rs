use std::path::Path;

use bun_native_plugin::{bun, define_bun_plugin, Result};
use oxc::{
    allocator::Allocator,
    ast::ast::{
        BindingRestElement, CallExpression, Expression, FormalParameterKind, ImportDeclaration,
        ImportDeclarationSpecifier, Program, Statement, TSTypeAnnotation,
        TSTypeParameterDeclaration,
    },
    codegen::{Codegen, CodegenOptions, LegalComment},
    parser::Parser,
    semantic::SemanticBuilder,
    span::{SourceType, Span},
};
use oxc_traverse::{traverse_mut, Traverse, TraverseCtx};

use crate::state::{
    CompilerError, CompilerState, Discovery, ImportName, MagicFunction, NapiCompilerDiagnostic,
    Target,
};

#[macro_use]
extern crate napi_derive;

pub mod route;
pub mod state;

struct DiscoveryTraverser<'a, 'b> {
    state: &'b mut CompilerState<'a>,
    declarations_to_add: Vec<Statement<'a>>,
}

impl<'a> DiscoveryTraverser<'a, '_> {
    fn create_empty_iife(&self, ctx: &mut TraverseCtx<'a>) -> Expression<'a> {
        Expression::ArrowFunctionExpression(oxc::allocator::Box::new_in(
            ctx.ast.arrow_function_expression::<Option<oxc::allocator::Box<'a, TSTypeParameterDeclaration<'a>>>,_,Option<oxc::allocator::Box<'a, TSTypeAnnotation<'a>>>,_>(
                Span::new(0, 0),
                false,
                false,
                None,
                ctx.ast
                    .formal_parameters::<Option<oxc::allocator::Box<'a, BindingRestElement<'a>>>>(
                        Span::new(0, 0),
                        FormalParameterKind::ArrowFormalParameters,
                        ctx.ast.vec(),
                        None,
                    ),
                None,
                ctx.ast.function_body(
                    Span::new(0, 0),
                    ctx.ast.vec(),
                    ctx.ast.vec(),
                ),
            ),
            ctx.ast.allocator,
        ))
    }
}

impl<'a> Traverse<'a> for DiscoveryTraverser<'a, '_> {
    fn exit_program(&mut self, node: &mut Program<'a>, _ctx: &mut TraverseCtx<'a>) {
        while let Some(declaration) = self.declarations_to_add.pop() {
            node.body.push(declaration);
        }
    }

    fn enter_import_declaration(
        &mut self,
        node: &mut ImportDeclaration<'a>,
        _: &mut TraverseCtx<'a>,
    ) {
        let import_source = node.source.value.as_str();
        if let Some(specifiers) = &node.specifiers {
            for specifier in specifiers {
                let name = match specifier {
                    ImportDeclarationSpecifier::ImportDefaultSpecifier(_) => ImportName::Default,
                    ImportDeclarationSpecifier::ImportNamespaceSpecifier(_) => ImportName::Default,
                    ImportDeclarationSpecifier::ImportSpecifier(name) => {
                        ImportName::Named(name.imported.name().as_str())
                    }
                };
                let magic_function = MagicFunction::from_import(import_source, name);

                let Some(magic_function) = magic_function else {
                    continue;
                };

                let symbol_id = specifier.local().symbol_id();

                self.state
                    .magic_function_table
                    .insert(symbol_id, magic_function);
            }
        }
    }

    fn exit_call_expression(&mut self, node: &mut CallExpression<'a>, ctx: &mut TraverseCtx<'a>) {
        let callee = &node.callee;

        if let Expression::Identifier(ident) = callee {
            let ref_id = ident.reference_id();
            let Some(symbol) = ctx.scoping.scoping().get_reference(ref_id).symbol_id() else {
                return;
            };
            let Some(magic_function) = self.state.magic_function_table.get(&symbol) else {
                return;
            };
            match magic_function {
                MagicFunction::Route => {
                    self.handle_route_function(node, ctx);
                }
            }
        }
    }
}

#[napi(object)]
pub struct CompileResult {
    pub source: String,
    pub discoveries: Vec<Discovery>,
    pub diagnostics: Vec<NapiCompilerDiagnostic>,
}

fn compile(source: &str, language: SourceType) -> CompileResult {
    let allocator = Allocator::new();
    let parser = Parser::new(&allocator, source, language);
    let parsed = parser.parse();
    let mut program = parsed.program;
    let scoping = SemanticBuilder::new()
        .build(&program)
        .semantic
        .into_scoping();
    let mut state = CompilerState::new(Target::Client, source);

    for error in parsed.errors {
        state.add_error(CompilerError::OxcParsingError { data: error });
    }

    traverse_mut(
        &mut DiscoveryTraverser {
            state: &mut state,
            declarations_to_add: vec![],
        },
        &allocator,
        &mut program,
        scoping,
    );

    let source_map_path = Path::new("[smp]");

    let codegen = Codegen::new()
        .with_source_text(source)
        .with_options(CodegenOptions {
            single_quote: true,
            minify: true,
            comments: true,
            annotation_comments: true,
            legal_comments: LegalComment::Inline,
            source_map_path: Some(source_map_path.to_path_buf()),
        })
        .build(&program);

    let code = codegen.code;

    let code = code.replace(
        source_map_path.to_str().unwrap(),
        &codegen.map.unwrap().to_data_url(),
    );

    CompileResult {
        source: code,
        diagnostics: state.get_diagnostics(),
        discoveries: state.discoveries,
    }
}

#[napi]
#[allow(dead_code)]
fn compile_script(source: String, file_name: String) -> CompileResult {
    let language = SourceType::from_path(file_name).unwrap_or(SourceType::tsx());
    compile(&source, language)
}

define_bun_plugin!("replace-discovery-imports");

#[bun]
pub fn replace_discovery_imports(handle: &mut OnBeforeParse) -> Result<()> {
    let input_source_code = handle.input_source_code()?;

    let loader = handle.output_loader();

    let output = compile_script(input_source_code.to_string(), handle.path()?.to_string());

    handle.set_output_source_code(output.source, loader);

    Ok(())
}
