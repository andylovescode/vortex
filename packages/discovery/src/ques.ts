const discoveryCompilerCues = ["@vortexjs/wormhole/route", "@vortexjs/common"];

export function checkForCues(contents: string): boolean {
    return discoveryCompilerCues.some((cue) => contents.includes(cue));
}
