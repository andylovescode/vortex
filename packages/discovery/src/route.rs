use std::collections::HashMap;

use oxc::{
    allocator::CloneIn,
    ast::ast::{
        CallExpression, Declaration, Expression, Statement, TSTypeAnnotation,
        VariableDeclarationKind,
    },
    span::GetSpan,
};
use oxc_traverse::TraverseCtx;

use crate::{
    state::{CompilerError, Discovery, RouteFrame},
    DiscoveryTraverser,
};

impl<'a> DiscoveryTraverser<'a, '_> {
    pub fn create_export<'p>(
        &mut self,
        ctx: &mut TraverseCtx<'a>,
        to_export: &Expression<'p>,
    ) -> String {
        let span = to_export.span();
        let id = format!("discovery${}${}", span.start, span.end);

        let declarator = ctx.ast.variable_declarator(
            span,
            VariableDeclarationKind::Const,
            ctx.ast
                .binding_pattern::<Option<oxc::allocator::Box<TSTypeAnnotation>>>(
                    ctx.ast
                        .binding_pattern_kind_binding_identifier(span, ctx.ast.atom(&id)),
                    None,
                    false,
                ),
            Some(to_export.clone_in(ctx.ast.allocator)),
            false,
        );

        let var_decl = Declaration::VariableDeclaration(ctx.ast.alloc_variable_declaration(
            span,
            VariableDeclarationKind::Const,
            ctx.ast.vec_from_array([declarator]),
            false,
        ));

        self.declarations_to_add
            .push(Statement::ExportNamedDeclaration(
                ctx.ast
                    .plain_export_named_declaration_declaration(span, var_decl),
            ));

        id
    }

    pub fn handle_route_function(
        &mut self,
        node: &mut CallExpression<'a>,
        ctx: &mut TraverseCtx<'a>,
    ) {
        if node.arguments.len() != 2 {
            self.state
                .add_error(CompilerError::IncorrectNumberOfArguments {
                    span: node.span,
                    expected: 2,
                    found: node.arguments.len(),
                });
            return;
        }

        let route = self
            .state
            .get_literal_string(node.arguments.remove(0).to_expression());

        let opts_expr = node.arguments.remove(0);
        let mut opts: HashMap<String, &Expression> = HashMap::new();

        self.state
            .get_object_keys(opts_expr.to_expression(), &mut opts);

        for export in ["page", "layout"] {
            if opts.contains_key(export) {
                let export_id = self.create_export(ctx, opts.get(export).unwrap());

                node.arguments.clear();

                node.callee = self.create_empty_iife(ctx);

                let frame = match export {
                    "page" => RouteFrame::Page,
                    "layout" => RouteFrame::Layout,
                    _ => unreachable!(),
                };

                self.state.discoveries.push(Discovery::Route {
                    frame,
                    path: route.clone(),
                    export: export_id.clone(),
                });
            }
        }
    }
}
