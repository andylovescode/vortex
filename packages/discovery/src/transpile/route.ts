import { unwrap } from "@vortexjs/common";
import type { CallExpression } from "oxc-parser";
import { exportNode } from "../builders";
import {
	type CompilerState,
	getObjectKeys,
	getStringLiteralValue,
} from "../compiler";

export function handleRouteFunction(
	state: CompilerState,
	node: CallExpression,
) {
	state.transformer.remove(node);

	if (node.arguments.length !== 2) {
		state.errors.push({
			from: node.start,
			to: node.end,
			message: `expected 2 arguments, got ${node.arguments.length}`,
			hints: [],
		});
		return;
	}

	const options = getObjectKeys(state, unwrap(node.arguments[1]));
	const path = getStringLiteralValue(state, unwrap(node.arguments[0]));

	if (!options) return;
	if (!path) return;

	const keys = Object.keys(options);

	if (keys.length === 0) {
		state.errors.push({
			from: node.start,
			to: node.end,
			message: "expected options object to have at least one property",
			hints: [],
		});
		return;
	}

	const validKeys = ["page", "layout"] as const;

	for (const key of keys) {
		if (!validKeys.includes(key as any)) {
			state.errors.push({
				from: node.start,
				to: node.end,
				message: `invalid option key "${key}", expected one of ${validKeys.join(", ")}`,
				hints: [],
			});
		}
	}

	for (const key of validKeys) {
		const value = options[key];

		if (!value) continue;

		const exported = exportNode(state, value);

		state.discoveries.push({
			type: "route_frame",
			exported: exported,
			frameType: key,
			path,
		});
	}
}
