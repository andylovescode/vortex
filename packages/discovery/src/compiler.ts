import { eventSchema, nonConstantFastHash } from "@vortexjs/cache";
import { smolify, unsmolify } from "@vortexjs/common";
import type { PippinError } from "@vortexjs/pippin";
import { print } from "esrap";
import tsx from "esrap/languages/tsx";
import { type Expression, type Node, parseAsync } from "oxc-parser";
import type { Discovery, DiscoveryProps, DiscoveryTarget } from "./api";
import { resolveName, Scoper } from "./scoper";
import { Transformer } from "./transform";
import { handleAPIFunction } from "./transpile/api";
import { handleRouteFunction } from "./transpile/route";
import { inlineUnwrapCall } from "./transpile/unwrap";

export function getSpecialImport(module: string, imported: string) {
	if (
		module === "@vortexjs/wormhole/route" &&
		(imported === "route" || imported === "default")
	) {
		return "route";
	}
	if (module === "@vortexjs/wormhole/route" && imported === "query") {
		return "query";
	}
	if (module === "@vortexjs/wormhole/route" && imported === "mutation") {
		return "mutation";
	}
	if (module === "@vortexjs/common" && imported === "unwrap") {
		return "unwrap";
	}
}

export type SpecialImport = Exclude<
	ReturnType<typeof getSpecialImport>,
	undefined
>;

export interface CompilerState {
	transformer: Transformer;
	scoper: Scoper;
	errors: Omit<PippinError, "path">[];
	target: DiscoveryTarget;
	discoveries: Discovery[];
	clientEligible: boolean;
	serverEligible: boolean;
}

export function getObjectKeys(state: CompilerState, node: Node) {
	if (node.type !== "ObjectExpression") {
		state.errors.push({
			from: node.start,
			to: node.end,
			message: `expected an object, got ${node.type}`,
			hints: [],
		});
		return;
	}

	const result: Record<string, Expression> = {};

	for (const prop of node.properties) {
		if (prop.type !== "Property") {
			state.errors.push({
				from: prop.start,
				to: prop.end,
				message: `expected a property, got ${prop.type}`,
				hints: [],
			});
			continue;
		}

		if (prop.key.type !== "Identifier" && prop.key.type !== "Literal") {
			state.errors.push({
				from: prop.key.start,
				to: prop.key.end,
				message: `expected an identifier or string key, got ${prop.key.type}`,
				hints: [],
			});
			continue;
		}

		result[resolveName(prop.key)] = prop.value;
	}

	return result;
}

export function getStringLiteralValue(state: CompilerState, node: Node) {
	if (node.type === "Literal") {
		if (typeof node.value === "string") {
			return node.value;
		}
		state.errors.push({
			from: node.start,
			to: node.end,
			message: `expected a string literal, got ${node.type} with value ${node.value}`,
			hints: [],
		});
		return undefined;
	}
	if (node.type === "TemplateLiteral") {
		if (node.expressions.length > 0) {
			state.errors.push({
				from: node.start,
				to: node.end,
				message:
					"expected a string literal, got a template literal with expressions",
				hints: [],
			});
			return undefined;
		}
		return node.quasis.map((q) => q.value.cooked).join("");
	}
	state.errors.push({
		from: node.start,
		to: node.end,
		message: `expected a string literal, got ${node.type}`,
		hints: [],
	});
	return undefined;
}

type OutputData = Omit<
	Awaited<ReturnType<typeof discoveryCompile>>,
	`${"client" | "server"}Eligible`
>;

const discoveryTransformSpec = eventSchema("discovery:transform")
	.field<"clientEligible", boolean>("clientEligible")
	.field<"serverEligible", boolean>("serverEligible")
	.field("inputHash")
	.field<"output", OutputData>("output");

export async function discoveryCompile(props: DiscoveryProps): Promise<{
	source: string;
	errors: Omit<PippinError, "path">[];
	discoveries: Discovery[];
}> {
	const inputHash = nonConstantFastHash(
		JSON.stringify({
			jsx: props.jsx,
			ts: props.typescript,
			code: props.source,
		}),
	).toString(36);

	const output = props.cache?.query?.(
		discoveryTransformSpec,
		{
			inputHash,
			clientEligible: props.target === "client" ? true : undefined,
			serverEligible: props.target === "server" ? true : undefined,
		},
		"output",
	);

	if (output) {
		return {
			source: unsmolify(output.source),
			errors: output.errors,
			discoveries: output.discoveries,
		};
	}

	const parsed = (
		await parseAsync(props.fileName, props.source, {
			lang: props.typescript
				? props.jsx
					? "tsx"
					: "ts"
				: props.jsx
					? "jsx"
					: "js",
		})
	).program;

	const state: CompilerState = {
		transformer: new Transformer(),
		scoper: new Scoper(),
		errors: [],
		discoveries: [],
		target: props.target,
		clientEligible: true,
		serverEligible: true,
	};

	state.scoper.hook(state.transformer);

	state.transformer.onEnter("CallExpression", (node) => {
		const callee = node.callee;

		if (callee.type !== "Identifier") return;

		const identifier = callee.name;

		const value = state.scoper.resolve(identifier);

		if (!value || value.type !== "import") return;

		const specialImport = getSpecialImport(value.module, value.id);

		if (!specialImport) return;

		if (specialImport === "route") {
			handleRouteFunction(state, node);
		} else if (specialImport === "query" || specialImport === "mutation") {
			handleAPIFunction(state, node, specialImport);
		} else if (specialImport === "unwrap") {
			inlineUnwrapCall(state, node);
		}
	});

	state.transformer.transform(parsed);

	const { code, map } = print(parsed, tsx());

	props.cache?.supply?.(discoveryTransformSpec, {
		clientEligible: state.clientEligible,
		serverEligible: state.serverEligible,
		output: {
			source: smolify(code),
			errors: state.errors,
			discoveries: state.discoveries,
		},
		inputHash,
	});

	return {
		errors: state.errors,
		source: code,
		discoveries: state.discoveries,
	};
}
