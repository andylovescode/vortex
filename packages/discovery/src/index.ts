import type { PippinPlugin } from "@vortexjs/pippin";
import type { DiscoveryPluginProps, DiscoveryTarget } from "./api";
import { discoveryCompile } from "./compiler";
import { eventSchema, nonConstantFastHash } from "@vortexjs/cache";
import { hash } from "@vortexjs/common";
import { checkForCues } from "./ques";

export * from "./api";
export { discoveryCompile } from "./compiler";

export function pippinPluginDiscovery(
    opts: DiscoveryPluginProps,
): PippinPlugin {
    return {
        name: "Discovery",
        description: "Compiler to power ergonomic framework APIs",
        transformers: [
            {
                async transform(props) {
                    if (props.format.type !== "ecma") {
                        return;
                    }

                    if (!checkForCues(props.source)) {
                        return;
                    }

                    const { source, errors } = await discoveryCompile({
                        ...props,
                        target: opts.target,
                        jsx: props.format.jsx,
                        typescript: props.format.typescript,
                        fileName: props.path,
                        cache: props.cache
                    });

                    for (const error of errors) {
                        await props.logError(error);
                    }

                    return {
                        source,
                        format: props.format,
                    };
                },
            },
        ],
    };
}

export * from "./ques";
