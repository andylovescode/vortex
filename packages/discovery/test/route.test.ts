import { expect, test } from "bun:test";
import { discoveryCompile } from "../src";

test("client output", async () => {
	expect(
		await discoveryCompile({
			fileName: "test.wormhole",
			source: `
			import route from "@vortexjs/wormhole/route";

route(
    "/",
    {
        page() {
            return (
                <>
                    <h1 class="text-4xl font-bold">
                        Welcome to Wormhole, {Object.entries(globalThis).length}
                    </h1>
                    <p>
                        This is an example app, go to the{" "}
                        <a href="/docs/tada">docs</a>
                    </p>
                </>
            );
        },
        layout({ children }) {
            return (
                <>
                    <head>
                        <title>Wormhole Example</title>
                    </head>
                    <body>{children}</body>
                </>
            );
        },
    },
);

route(
    "/docs/[page]",
    {
        page({ page }) {
            return (
                <>
                    <h1>Documentation for {page}</h1>
                    <p>This is the documentation page for {page}.</p>
                </>
            );
        },
    }
);
        `,
			jsx: true,
			typescript: true,
			target: "client",
		}),
	).toMatchSnapshot();

	expect(
		await discoveryCompile({
			fileName: "test.wormhole",
			source: `import { query } from "@vortexjs/wormhole/route";

	const a = query("/api/data", {
		impl: "data",
		schema: "abc",
		method: "DELETE",
	});

	a(); `,
			jsx: true,
			typescript: true,
			target: "client",
		}),
	).toMatchSnapshot();
});
