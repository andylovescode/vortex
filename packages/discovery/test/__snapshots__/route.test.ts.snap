// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`client output 1`] = `
{
  "discoveries": [
    {
      "exported": "$d_0",
      "frameType": "page",
      "path": "/",
      "type": "route_frame",
    },
    {
      "exported": "$d_1",
      "frameType": "layout",
      "path": "/",
      "type": "route_frame",
    },
    {
      "exported": "$d_2",
      "frameType": "page",
      "path": "/docs/[page]",
      "type": "route_frame",
    },
  ],
  "errors": [],
  "source": 
"import route from "@vortexjs/wormhole/route";

export const $d_0 = function () {
	return (<>
                    <h1 class="text-4xl font-bold">
                        Welcome to Wormhole, {Object.entries(globalThis).length}
                    </h1>
                    <p>
                        This is an example app, go to the{" "}
                        <a href="/docs/tada">docs</a>
                    </p>
                </>);
};

export const $d_1 = function ({ children }) {
	return (<>
                    <head>
                        <title>Wormhole Example</title>
                    </head>
                    <body>{children}</body>
                </>);
};

export const $d_2 = function ({ page }) {
	return (<>
                    <h1>Documentation for {page}</h1>
                    <p>This is the documentation page for {page}.</p>
                </>);
};"
,
}
`;

exports[`client output 2`] = `
{
  "discoveries": [],
  "errors": [],
  "source": 
"import { query } from "@vortexjs/wormhole/route";

const a = $i_0({ schema: "abc", pathname: '/api/data', method: 'DELETE' });

a();

import { INTERNAL_client_query as $i_0 } from '@vortexjs/wormhole';"
,
}
`;
